packages/ragbits-core/src/ragbits/core/pipelines/pipeline.py

```diff
@@ -0,0 +1,106 @@
+from abc import ABC, abstractmethod
+from types import UnionType
+from typing import Generic, TypeVar, Union, cast, get_type_hints
+
+from generics import get_filled_type
+from typing_utils import Type, issubtype
+
+InputT = TypeVar("InputT")
+OutputT = TypeVar("OutputT")
+
+
+class Step(ABC, Generic[InputT, OutputT]):
+    """
+    A step in a pipeline that processes input data and returns output data
+    """
+
+    @abstractmethod
+    async def run(self, /, input_data: InputT) -> OutputT:
+        """
+        Run the step with the given input data
+        """
+
+
+class Pipeline(Step[InputT, OutputT]):
+    """
+    A pipeline that consists of multiple steps that process input data and return output data
+    """
+
+    def __init__(self, *steps: Step) -> None:
+        self.steps = steps
+        self._validate_intrastep_types()
+
+    @staticmethod
+    def _issubtype(needed_type: Type, actual_type: Type) -> bool:
+        """
+        Check if the actual type is a subtype of the needed type.
+        """
+        # Covert new union type syntax to Union objects, as typing_utils.issubtype
+        # does not support the new syntax
+        if isinstance(needed_type, UnionType):
+            needed_type = Union[tuple(needed_type.__args__)]  # noqa: UP007
+        if isinstance(actual_type, UnionType):
+            actual_type = Union[tuple(actual_type.__args__)]  # noqa: UP007
+        return issubtype(needed_type, actual_type)
```

Comment:
`issubtype` comes from an external library (`typing_utils`) and unfortunately it is on the library's author to implement all the logic behind different typing hints in Python.

Which means that it usually works, but in some cases (new additions to the typing system, less know types, edge cases, generics) may not. For example, the library don't support new pipe syntax for unions - that's why I had to convert from new to old syntax above.

I see this as a problem, because if we go with this implementation, it will _look_ like Steps and Pipelines can be typed with any typing hints available, and usually it indeed work. But for some more advanced hints it will just not.